### PR TITLE
[sweet api][kotlin] Cache record converters

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Small optimization ;) 

# How

Cached created record converters and reuse them later. 

# Test Plan

- unit tests

### CODE REVIEW

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt [hunk 1]
```diff
@@ -68,9 +69,15 @@ object TypeConverterProviderImpl : TypeConverterProvider {
       return EnumTypeConverter(kClass as KClass<Enum<*>>, type.isMarkedNullable)
     }
 
-    // TODO(@lukmccall): cache record type converters
+    val maybeConverter = cachedRecordConverters[kClass]
+    if (maybeConverter != null) {
+      return maybeConverter
+    }
```
Comment (tsapeta) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt:75]: `maybe` sounds kinda weird here 🙈 Why not `cached` instead?

```suggestion
    val cachedConverter = cachedRecordConverters[kClass]
    if (cachedConverter != null) {
      return cachedConverter
    }
```

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt [hunk 2]
```diff
@@ -68,9 +69,15 @@ object TypeConverterProviderImpl : TypeConverterProvider {
...
+
     if (kClass.isSubclassOf(Record::class)) {
-      return RecordTypeConverter<Record>(this, type)
+      val converter = RecordTypeConverter<Record>(this, type)
```
Comment (tsapeta) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt:78]: Is the creation of the record type converter so expensive to be worth caching?

Comment (lukmccall) [reply packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt:78]: Nope, but we can easily reuse converters. For instance, if you have two methods in the module which uses the same `Record`, we don't have to create a separate object for them. In that case, we save a little bit of memory. Currently, records converters hold all the needed information to create a Kotlin object. In that way, we have an extremely fast runtime with not that huge memory overhead. IMO, it's worth caching.

**Review by barthap — APPROVED** [review]

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt (hunk 2 above)
Comment (bbarthec) [reply packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt:78]: TBH, I would like to see some benchmarks 😅 
I get that it's super awesome when we do optimise the code, but I don't see the reason for doing so if the more naive way is performant enough 🤔

**Review by bbarthec — COMMENTED** [review]
Wouldn't that be easier if we push the caching layer down to the constructor?
I guess we can have a Factory that would provide us with the fresh converter or reuse cached instance 👀 
Using such mechanism is more scalable I guess 🤔

**Comment by lukmccall** [comment]
> TBH, I would like to see some benchmarks 😅
I get that it's super awesome when we do optimise the code, but I don't see the reason for doing so if the more naive way is performant enough 🤔

TBH, I don't see what should I benchmark here. This improvement isn't huge but in my opinion, adding caching here is a more natural way of implementing that. Let me explain that. Converters are already stored in the `AnyType` class. So this code doesn't use more memory - it only adds addition in the memory needed to store a record converters map. Also, we already cached simple converters (for int, float, etc.), because we know those objects will be used in multiple places, and therefore, I don't see a reason why not to cache record converters too. Recreating them is just a waste of time. All of them will look the same and they are stateless. So, IMO, they should be cached. 

> Wouldn't that be easier if we push the caching layer down to the constructor?

I'm not sure to which constructor you are referring. But anyway, I don't see a value in making it more scalable here by adding more complex structures. It's very unlikely that we ever will change `TypeConverter Provider` in the future.

**Review by bbarthec — APPROVED** [review]

**Pull request merged**